### PR TITLE
Improve enable_extended_results_for_datetime_functions option to return results of type DateTime64

### PR DESCRIPTION
**Pull request by rvasin**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve option enable_extended_results_for_datetime_functions  to return results of type DateTime64 for functions toStartOfDay, toStartOfHour, toStartOfFifteenMinutes, toStartOfTenMinutes,  toStartOfFiveMinutes, toStartOfMinute and timeSlot.

Continuation of #41214

### CODE REVIEW

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+template <typename Transform>
+class FunctionDateOrDateTimeToDateTimeOrDateTime64 : public IFunctionDateOrDateTime<Transform>, WithContext
+{
+public:
+    const bool enable_extended_results_for_datetime_functions = false;
```
Comment (rschu1ze) [inline src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:16]: Minor: `enable_extended_results_for_datetime_functions` could be in the private section (same applies to "FunctionDateOrDateTimeToDateOrDate32.h")

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+template <typename Transform>
+class FunctionDateOrDateTimeToDateTimeOrDateTime64 : public IFunctionDateOrDateTime<Transform>, WithContext
```
Comment (rschu1ze) [inline src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:13]: Minor: I don't think it is necessary to inherit from `WithContext` (same applies to "FunctionDateOrDateTimeToDateOrDate32.h")

File: src/Functions/timeSlot.cpp [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
-#include <Functions/FunctionDateOrDateTimeToSomething.h>
+#include <Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h>
 #include <DataTypes/DataTypesNumber.h>
 
 
 namespace DB
 {
 
-using FunctionTimeSlot = FunctionDateOrDateTimeToSomething<DataTypeDateTime, TimeSlotImpl>;
+using FunctionTimeSlot = FunctionDateOrDateTimeToDateTimeOrDateTime64<TimeSlotImpl>;
```
Comment (rschu1ze) [inline src/Functions/timeSlot.cpp:11]: (Minor: Could mention function `timeSlot` in the PR description.)

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            return std::make_shared<DataTypeDateTime64>(scale, time_zone);
+        }
+        else
+            return std::make_shared<DataTypeDateTime>(time_zone);
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
+    {
+        const IDataType * from_type = arguments[0].type.get();
+        WhichDataType which(from_type);
```
Comment (rschu1ze) [inline src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:62]: \+ newline

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        if (which.isDate())
+            return DateTimeTransformImpl<DataTypeDate, DataTypeDateTime, Transform>::execute(arguments, result_type, input_rows_count);
+        else if (which.isDate32())
+            if (enable_extended_results_for_datetime_functions)
+                return DateTimeTransformImpl<DataTypeDate32, DataTypeDateTime64, Transform, /*is_extended_result*/ true>::execute(arguments, result_type, input_rows_count);
+            else
+                return DateTimeTransformImpl<DataTypeDate32, DataTypeDateTime, Transform>::execute(arguments, result_type, input_rows_count);
+        else
```
Comment (rschu1ze) [inline src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:70]: funny else/if indentation

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            const auto scale = static_cast<const DataTypeDateTime64 *>(from_type)->getScale();
+
+            const TransformDateTime64<Transform> transformer(scale);
+            if (enable_extended_results_for_datetime_functions)
+                return DateTimeTransformImpl<DataTypeDateTime64, DataTypeDateTime64, decltype(transformer), /*is_extended_result*/ true>::execute(arguments, result_type, input_rows_count, transformer);
+            else
+                return DateTimeTransformImpl<DataTypeDateTime64, DataTypeDateTime, decltype(transformer)>::execute(arguments, result_type, input_rows_count, transformer);
+        }
+        else
+            throw Exception("Illegal type " + arguments[0].type->getName() + " of argument of function " + this->getName(),
```
Comment (rschu1ze) [inline src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:84]: (Bonus points for using C++20 `format()` syntax for `throw`:

```cpp
throw Exception(
    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
    "Illegal type {} of argument of function {}",
    arguments[0].type->getName(), this->getName());
```

here and in other places.)

File: src/Functions/DateTimeTransforms.h [hunk 7]
```diff
@@ -46,6 +46,11 @@ namespace ErrorCodes
         throw Exception("Illegal type DateTime of argument for function " + std::string(name), ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
     }
 
+    static inline Int64 date32IsNotSupported(const char * name)
```
Comment (rschu1ze) [inline src/Functions/DateTimeTransforms.h:49]: This and above methods could be `[[noreturn]]`. Bonus points for moving their implementation into a to-be-created source file.

File: src/Functions/DateTimeTransforms.h [hunk 8]
```diff
@@ -136,6 +140,14 @@ struct ToStartOfDayImpl
         auto date_time = time_zone.fromDayNum(ExtendedDayNum(d));
         return date_time < 0xffffffff ? date_time : time_zone.toDate(0xffffffff);
     }
+    static inline DecimalUtils::DecimalComponents<DateTime64> execute_extended_result(const DecimalUtils::DecimalComponents<DateTime64> & t, const DateLUTImpl & time_zone)
```
Comment (rschu1ze) [inline src/Functions/DateTimeTransforms.h:143]: Costmetic: My local clang format complains about case style. Should have been `executeExtendedResult`.

File: src/Functions/DateTimeTransforms.h [hunk 9]
```diff
@@ -624,6 +667,16 @@ struct TimeSlotImpl
...
     }
 
+    static inline DecimalUtils::DecimalComponents<DateTime64> execute_extended_result(const DecimalUtils::DecimalComponents<DateTime64> & t, const DateLUTImpl & time_zone)
+    {
+       return {time_zone.toStartOfMinuteInterval(t.whole, 30), 0};
+    }
+
+    static inline Int64 execute_extended_result(Int32, const DateLUTImpl &)
+    {
+        return date32IsNotSupported(name);
```
Comment (rschu1ze) [inline src/Functions/DateTimeTransforms.h:677]: Looks a bit weird that the `*IsNotSupported()` methods return an integer. They ideally return `void` and we could just call `date32IsNotSupported()` here. But perhaps it is because we return Int64 in l. 675 and the compiler gets confused otherwise.

File: src/Functions/DateTimeTransforms.h [hunk 10]
```diff
@@ -624,6 +667,16 @@ struct TimeSlotImpl
         return dateIsNotSupported(name);
     }
 
+    static inline DecimalUtils::DecimalComponents<DateTime64> execute_extended_result(const DecimalUtils::DecimalComponents<DateTime64> & t, const DateLUTImpl & time_zone)
+    {
+       return {time_zone.toStartOfMinuteInterval(t.whole, 30), 0};
```
Comment (rschu1ze) [inline src/Functions/DateTimeTransforms.h:672]: Maybe I miss something but should l. 672 not correspond to l. 652?

File: src/Common/DateLUTImpl.h [hunk 11]
```diff
@@ -1030,7 +1030,7 @@ class DateLUTImpl
     template <typename DateOrTime>
     DateOrTime toStartOfMinuteInterval(DateOrTime t, UInt64 minutes) const
     {
-        UInt64 divisor = 60 * minutes;
+        Int64 divisor = 60 * minutes;
```
Comment (rschu1ze) [inline src/Common/DateLUTImpl.h:1033]: Hm, why?

File: docs/en/operations/settings/settings.md [hunk 12]
```diff
@@ -3147,12 +3147,14 @@ Result:
...
+-   `DateTime64` with extended range (compared to type `DateTime`) for functions [toStartOfDay](../../sql-reference/functions/date-time-functions.md#tostartofday), [toStartOfHour](../../sql-reference/functions/date-time-functions.md#tostartofhour), [toStartOfMinute](../../sql-reference/functions/date-time-functions.md#tostartofminute), [toStartOfFiveMinutes](../../sql-reference/functions/date-time-functions.md#tostartoffiveminutes), [toStartOfTenMinutes](../../sql-reference/functions/date-time-functions.md#tostartoftenminutes), [toStartOfFifteenMinutes](../../sql-reference/functions/date-time-functions.md#tostartoffifteenminutes) and [timeSlot](../../sql-reference/functions/date-time-functions.md#timeslot).
 
 Possible values:
 
--   0 — Functions return `Date` for all types of arguments.
--   1 — Functions return `Date32` for `Date32` or `DateTime64` arguments and `Date` otherwise.
+-   0 — Functions return `Date` or `DateTime` for all types of arguments.
```
Comment (rschu1ze) [inline docs/en/operations/settings/settings.md:3156]: Would be cool to also update the "note" box in https://github.com/ClickHouse/ClickHouse/pull/41214/files#diff-c1b9883c71837d78852560daed73fe8588bf02cd2d5ba619144d8793facc785a

**Review by rschu1ze — COMMENTED** [review]
Looks overall good to me. Attaching some smaller comments.

File: src/Common/DateLUTImpl.h (hunk 11 above)
Comment (ilejn) [reply src/Common/DateLUTImpl.h:1033]: Because unsigned is promoted in C and C++.
Consider 'Time res = date + (t - date) / divisor * divisor' , for example.

File: src/Functions/timeSlot.cpp (hunk 3 above)
Comment (rvasin) [reply src/Functions/timeSlot.cpp:11]: Done

File: src/Functions/DateTimeTransforms.h (hunk 9 above)
Comment (ilejn) [reply src/Functions/DateTimeTransforms.h:677]: [[noreturn]] probably helps

File: src/Common/DateLUTImpl.h (hunk 11 above)
Comment (rvasin) [reply src/Common/DateLUTImpl.h:1033]: We found the bug that functions like toStartOfFiveMinutes (which are using toStartOfMinuteInterval) did work well for 1970 and later dates. But they did not work at all for dates before 1970. This is because t is negative. And division like:
negative t by unsigned divisor produced a "garbage".

Consider the example:
```
int main()
{
    unsigned int divisor = 60;
    int t = -90;
    cout << t / divisor << endl;
    return 0;
}
```
It returns 71582786.

But division of signed int by signed int works fine:
```
int main()
{
    int divisor = 60;
    int t = -90;
    cout << t / divisor << endl;
    return 0;
}
```
It returns -1.

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        const IDataType * from_type = arguments[0].type.get();
+        WhichDataType which(from_type);
+
+        std::string time_zone = extractTimeZoneNameFromFunctionArguments(arguments, 1, 0);
+
+        /// If the time zone is specified but empty, throw an exception.
+        /// only validate the time_zone part if the number of arguments is 2.
+        if (arguments.size() == 2 && time_zone.empty())
+            throw Exception(
+                "Function " + this->getName() + " supports a 2nd argument (optional) that must be non-empty and be a valid time zone",
```
Comment (ilejn) [inline src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:42]: Are you sure that 'must be non-empty' makes sense?
Is not it better " ... (optional). If specified must be a valid time zone" or just "... (optional) that must be a valid time zone" ?

File: tests/queries/0_stateless/02403_enable_extended_results_for_datetime_functions.reference [hunk 14]
```diff
@@ -54,3 +70,19 @@ toLastDayOfMonth;toDate32;false	1970-01-01
 type;toLastDayOfMonth;toDate32;false	Date
 toLastDayOfMonth;toDateTime64;false	1970-01-01
 type;toLastDayOfMonth;toDateTime64;false	Date
+toStartOfDay;toDateTime64;false	1970-01-01 00:00:00
+type;toStartOfDay;toDateTime64;false	DateTime
+toStartOfHour;toDateTime64;false	1970-01-01 00:00:00
+type;toStartOfHour;toDateTime64;false	DateTime
+toStartOfMinute;toDateTime64;false	1970-01-01 00:00:00
+type;toStartOfMinute;toDateTime64;false	DateTime
+toStartOfFiveMinutes;toDateTime64;false	2056-03-09 16:48:16
```
Comment (ilejn) [inline tests/queries/0_stateless/02403_enable_extended_results_for_datetime_functions.reference:79]: Not sure that I understand why '16:48:16' is expected for toStartOfFiveMinutes

Comment (rvasin) [reply tests/queries/0_stateless/02403_enable_extended_results_for_datetime_functions.reference:79]: This is because "narrowing" is not implemented for toStartOfFiveMinutes, toStartOfTenMinutes, toStartOfFifteenMinutes and timeSlot function when the option enable_extended_results_for_datetime_functions is OFF. When the option is ON it works fine. @rschu1ze What do you think about it? Is it OK for current PR? Or should add narrowing for these functions in current PR for the case when the option is OFF.

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h (hunk 4 above)
Comment (rvasin) [reply src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:62]: Done

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h (hunk 2 above)
Comment (rvasin) [reply src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:13]: It's not minor. You are right: `WithContext` is not needed. I removed it from both places.

File: src/Common/DateLUTImpl.h (hunk 11 above)
Comment (rschu1ze) [reply src/Common/DateLUTImpl.h:1033]: I see, thanks. No objections then. It would be nice though if we had tests for the edge cases, i.e. input values close to the min/max date/time range.

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h (hunk 13 above)
Comment (rschu1ze) [reply src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:42]: I would second `... (optional) that must be a valid time zone`.

File: tests/queries/0_stateless/02403_enable_extended_results_for_datetime_functions.reference (hunk 14 above)
Comment (rschu1ze) [reply tests/queries/0_stateless/02403_enable_extended_results_for_datetime_functions.reference:79]: From a correctness POV, the current behavior is not great. But I think that the compatibility POV is more important here, so we better don't break existing behavior. At some point in future, `enable_extended_results_for_datetime_functions` will be `ON` by default, and we can stop worrying about this issue.

File: src/Functions/DateTimeTransforms.h (hunk 8 above)
Comment (rvasin) [reply src/Functions/DateTimeTransforms.h:143]: Yes, it's even in [recommendations](https://clickhouse.com/docs/en/development/style/#names):

> For the names of functions (methods), use camelCase beginning with a lowercase letter.

So I renamed the functions as needed.

File: tests/queries/0_stateless/02403_enable_extended_results_for_datetime_functions.sql.j2 [hunk 15]
```diff
@@ -5,5 +5,10 @@ SELECT 'type;{{ date_fun }};toDate32;{{ option_value }}', toTypeName({{ date_fun
 SELECT '{{ date_fun }};toDateTime64;{{ option_value }}', {{ date_fun }}(toDateTime64('1920-02-02 10:20:30', 3)) SETTINGS enable_extended_results_for_datetime_functions = {{ option_value }};
 SELECT 'type;{{ date_fun }};toDateTime64;{{ option_value }}', toTypeName({{ date_fun }}(toDateTime64('1920-02-02 10:20:30', 3))) SETTINGS enable_extended_results_for_datetime_functions = {{ option_value }};
 {% endfor -%}
+{% for date_fun in ['toStartOfDay', 'toStartOfHour', 'toStartOfMinute', 'toStartOfFiveMinutes', 'toStartOfTenMinutes', 'toStartOfFifteenMinutes', 'timeSlot'] -%}
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02403_enable_extended_results_for_datetime_functions.sql.j2:8]: I am not sure if such tests exist already, but it would be cool to have a few tests for "boundary cases", i.e. the absolute min/max values for each input value range. If it is inconvenient to integrate them into this test, a separate SQL test is also good.

Comment (rvasin) [reply tests/queries/0_stateless/02403_enable_extended_results_for_datetime_functions.sql.j2:8]: I am also not sure if there are tests for  "boundary cases" but after code review I found several boundary issues in about 5-6 date/time functions. I am going to create a mini PR to fix these boundary issues. So PR will surely have the boundary test.

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h (hunk 1 above)
Comment (rvasin) [reply src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:16]: Done

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h (hunk 5 above)
Comment (rvasin) [reply src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:70]: Done

File: src/Functions/DateTimeTransforms.h (hunk 7 above)
Comment (rvasin) [reply src/Functions/DateTimeTransforms.h:49]: I put no return this way: 
`[[noreturn]] static inline Int64 date32IsNotSupported(const char * name)`
for all 3 methods.

File: tests/queries/0_stateless/02403_enable_extended_results_for_datetime_functions.reference (hunk 14 above)
Comment (rvasin) [reply tests/queries/0_stateless/02403_enable_extended_results_for_datetime_functions.reference:79]: OK. We will keep it as it is for now.

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h (hunk 13 above)
Comment (rvasin) [reply src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:43]: done

File: src/Functions/DateTimeTransforms.h (hunk 10 above)
Comment (rvasin) [reply src/Functions/DateTimeTransforms.h:672]: @rschu1ze This is one of the interesting points. We have:

```
struct TimeSlotImpl
{
    static inline UInt32 execute(const DecimalUtils::DecimalComponents<DateTime64> & t, const DateLUTImpl &)
    {
        return t.whole / 1800 * 1800;
    }
    
    static inline DecimalUtils::DecimalComponents<DateTime64> executeExtendedResult(const DecimalUtils::DecimalComponents<DateTime64> & t, const DateLUTImpl & time_zone)
    {
       return {time_zone.toStartOfMinuteInterval(t.whole, 30), 0};
    } 
```

So do you think that executeExtendedResult also return result without timezone? 
It should be something like this:
`return {t.whole / 1800 * 1800, 0};`

File: src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h (hunk 6 above)
Comment (rvasin) [reply src/Functions/FunctionDateOrDateTimeToDateTimeOrDateTime64.h:86]: Done

File: docs/en/operations/settings/settings.md (hunk 12 above)
Comment (rvasin) [reply docs/en/operations/settings/settings.md:3156]: Done. Please take a look on the updated next in the "note" box.

File: src/Functions/DateTimeTransforms.h (hunk 10 above)
Comment (rvasin) [reply src/Functions/DateTimeTransforms.h:672]: I am working on this now. The case is interesting. For example:
`select timeSlot(toDateTime('1979-12-31 10:35:30'), 'Europe/Paris') `
returns 1979-12-31 11:30:00 (with timezone).
while:
`select timeSlot(toDateTime('1979-12-31 10:35:30', 'Europe/Paris')) `
returns without timezone: 1979-12-31 10:30:00

Probably I will implement the same for DateTime64 argument with enable_extended_results_for_datetime_functions = true.

Comment (rvasin) [reply src/Functions/DateTimeTransforms.h]: OK. I implement timeSlot() using more similar approach:

```
    static inline DecimalUtils::DecimalComponents<DateTime64>  executeExtendedResult(const DecimalUtils::DecimalComponents<DateTime64> & t, const DateLUTImpl &)
    {
        if (likely(t.whole >= 0))
            return {t.whole / 1800 * 1800, 0};
        return {(t.whole + 1 - 1800) / 1800 * 1800, 0};
    }
```
For negative values (dates before 1970) we still need to do things like:` t.whole + 1 - 1800) / 1800 * 1800 ` to make correct rounding. Otherwise 1969-12-31 13:35:30 will be rounded up to 1969-12-31 14:00:00 (instead of correct  1969-12-31 13:30:00 )

File: src/Functions/TransformDateTime64.h [hunk 16]
```diff
@@ -89,20 +89,20 @@ class TransformDateTime64
 
 
     template <typename ... Args>
-    inline auto NO_SANITIZE_UNDEFINED execute_extended_result(const DateTime64 & t, Args && ... args) const
+    inline auto NO_SANITIZE_UNDEFINED executeExtendedResult(const DateTime64 & t, Args && ... args) const
```
Comment (rschu1ze) [inline src/Functions/TransformDateTime64.h:92]: It would be nice if you could apply the same change `execute_extended_result` --> `executeExtendedResult` also in Functions/CustomWeekTransforms.h, Functions/DateTimeTransforms.h and Functions/TransformDateTime64.h. (Feel free to open a new PR for such cosmetic cleanups. Putting that into this PR is also fine.)

File: docs/en/operations/settings/settings.md (hunk 12 above)
Comment (rschu1ze) [reply docs/en/operations/settings/settings.md:3156]: Sounds good, overall. I have added a small comment about the order of mentioned functions.

File: src/Functions/TransformDateTime64.h (hunk 16 above)
Comment (rvasin) [reply src/Functions/TransformDateTime64.h:92]: Just to be sure I verified it today: 
I see that yesterday I already renamed execute_extended_result --> executeExtendedResult in 
CustomWeekTransforms.h, DateTimeTransforms.h and TransformDateTime64.h.

File: src/Functions/DateTimeTransforms.h (hunk 10 above)
Comment (ilejn) [reply src/Functions/DateTimeTransforms.h]: @rvasin , You've checked that it works for non default scale, right?

File: docs/en/sql-reference/functions/date-time-functions.md [hunk 17]
```diff
@@ -268,15 +268,17 @@ Result:
...
   *  If `toLastDayOfMonth` is called with an argument greater then `2149-05-31`, the result will be calculated from the argument `2149-05-31` instead.
-* `enable_extended_results_for_datetime_functions = 1`: Functions `toStartOfYear`, `toStartOfISOYear`, `toStartOfQuarter`, `toStartOfMonth`, `toStartOfWeek`, `toLastDayOfMonth`, `toMonday` return `Date` or `DateTime` if their argument is a `Date` or `DateTime`, and they return `Date32` or `DateTime64` if their argument is a `Date32` or `DateTime64`.
+* `enable_extended_results_for_datetime_functions = 1`:
+  * Functions `toStartOfYear`, `toStartOfISOYear`, `toStartOfQuarter`, `toStartOfMonth`, `toStartOfWeek`, `toLastDayOfMonth`, `toMonday` return `Date` or `DateTime` if their argument is a `Date` or `DateTime`, and they return `Date32` or `DateTime64` if their argument is a `Date32` or `DateTime64`.
+  * Functions `toStartOfDay`, `toStartOfHour`, `toStartOfMinute`, `toStartOfFiveMinutes`, `toStartOfTenMinutes`, `toStartOfFifteenMinutes`, `timeSlot` return `DateTime` if their argument is a `Date` or `DateTime`, and they return `DateTime64` if their argument is a `Date32` or `DateTime64`.
```
[the remark this answers is not in the document]
Comment (rvasin) [reply docs/en/sql-reference/functions/date-time-functions.md]: Done and pushed my changes. I also updates the list of functions in changes log text.

File: src/Functions/DateTimeTransforms.h (hunk 10 above)
Comment (ilejn) [reply src/Functions/DateTimeTransforms.h]: All this business with "+1 - 1800" seems correct. The shame is I cannot think up a helpful comment.

Comment (rvasin) [reply src/Functions/DateTimeTransforms.h]: Yes, I have verified: it works fine with non-default scale:
`select timeSlot(toDateTime64('1969-12-31 10:35:30.75012', 5), 'Europe/Paris')  SETTINGS enable_extended_results_for_datetime_functions = true;`
Returns: 1969-12-31 11:30:00.00000

File: src/Functions/TransformDateTime64.h (hunk 16 above)
Comment (rschu1ze) [reply src/Functions/TransformDateTime64.h:92]: Good, thanks.

File: src/Functions/DateTimeTransforms.h (hunk 10 above)
Comment (rschu1ze) [reply src/Functions/DateTimeTransforms.h]: I am good with this implementation.

**Review by rschu1ze — APPROVED** [review]
